libethereum/ExtVM.cpp

```diff
@@ -95,30 +95,94 @@ void go(unsigned _depth, Executive& _e, OnOpFunc const& _onOp)
 
 bool ExtVM::call(CallParameters& _p)
 {
-	Executive e(m_s, envInfo(), m_sealEngine, depth + 1);
+	m_successfulCalls.emplace_back(m_s, envInfo(), m_sealEngine, depth + 1);
+	auto& e = m_successfulCalls.back();
 	if (!e.call(_p, gasPrice, origin))
 	{
 		go(depth, e, _p.onOp);
 		e.accrueSubState(sub);
 	}
 	_p.gas = e.gas();
 
-	return !e.excepted();
+	if (e.excepted())
+	{
+		m_successfulCalls.pop_back();
+		return false;
+	}
+	return true;
 }
 
 size_t ExtVM::codeSizeAt(dev::Address _a)
 {
 	return m_s.codeSize(_a);
 }
 
+void ExtVM::setStore(u256 _n, u256 _v)
+{
+	if (!m_origStorage.count(_n))
+	{
+		m_origStorage.emplace(_n, store(_n));
+//		clog(ExecutiveWarnChannel) << "ORIG STORAGE " << myAddress << _n << _v;
+	}
+	m_s.setStorage(myAddress, _n, _v);
+}
+
 h160 ExtVM::create(u256 _endowment, u256& io_gas, bytesConstRef _code, OnOpFunc const& _onOp)
 {
-	Executive e(m_s, envInfo(), m_sealEngine, depth + 1);
+	// Every CREATE increases this account nonce, no matter if it succeeds.
+	++m_nonceInc;
+
+	m_successfulCalls.emplace_back(m_s, envInfo(), m_sealEngine, depth + 1);
+	auto& e = m_successfulCalls.back();
 	if (!e.create(myAddress, _endowment, gasPrice, io_gas, _code, origin))
 	{
 		go(depth, e, _onOp);
 		e.accrueSubState(sub);
 	}
 	io_gas = e.gas();
+	if (!e.newAddress())
+		m_successfulCalls.pop_back();
 	return e.newAddress();
 }
+
+void ExtVM::suicide(Address _a)
+{
+	// FIXME: What if _a is 0?
+	if (!m_s.isTouched(_a))
+	{
+//		clog(ExecutiveWarnChannel) << "Log SELFDESTRUCT  " << myAddress << _a;
+		m_selfdestructBeneficiary = _a;
+	}
+	m_s.addBalance(_a, m_s.balance(myAddress));
+	m_s.subBalance(myAddress, m_s.balance(myAddress));
+	ExtVMFace::suicide(_a);
+}
+
+void ExtVM::revert()
+{
+//	clog(ExecutiveWarnChannel) << "Reverting " << myAddress;
+	for (auto it = m_successfulCalls.rbegin(); it != m_successfulCalls.rend(); ++it)
+		it->revert();
+
+	// Restore original storage for this account. The order does not matter.
+//	clog(ExecutiveWarnChannel) << "Reverting storage " << myAddress;
+	for (auto& item: m_origStorage)
+	{
+		m_s.setStorage(myAddress, item.first, item.second);
+//		clog(ExecutiveWarnChannel) << "REVERT STORAGE " << myAddress << item.first << item.second;
+	}
+
+	// Revert nonce if dumped.
+	if (m_nonceInc > 0)
+		m_s.setNonce(myAddress, m_s.getNonce(myAddress) - m_nonceInc);
+
+	if (m_selfdestructBeneficiary)
+	{
+//		clog(ExecutiveWarnChannel) << "REVERT SELFDESTRUCT touch " << myAddress << m_selfdestructBeneficiary;
+		m_s.untouch(m_selfdestructBeneficiary);
+	}
+
```

Comment:
Good point. I think the contract of usage of `revert()` should be use only once. We can make second `revert()` doing nothing, but that is too defensive in my opinion and need additional care.